core/src/main/java/org/mule/transformer/types/TypedValue.java

```diff
@@ -9,31 +9,110 @@
 
 import org.mule.api.transformer.DataType;
 
+import java.io.NotSerializableException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
 import java.io.Serializable;
 
 /**
- *  Maintains a value that has an associated {@link DataType}
+ * Maintains a value that has an associated {@link DataType}
  */
-public class TypedValue implements Serializable
+public class TypedValue<T> implements Serializable
 {
 
-    private final Object value;
+    private static final long serialVersionUID = -2533879516750283994L;
 
-    private final DataType<?> dataType;
+    private transient T value;
 
-    public TypedValue(Object value, DataType<?> dataType)
+    // TODO MULE-9279 Make 'final' once MuleMessage is immutable
+    // TODO MULE-9288 DataType should be immutable
+    private DataType<T> dataType;
+
+    public TypedValue(T value, DataType<T> dataType)
     {
         this.value = value;
-        this.dataType = dataType;
+        if (dataType == null)
+        {
+            this.dataType = DataTypeFactory.create((Class<T>) value.getClass());
+        }
+        else
+        {
+            this.dataType = dataType;
+        }
     }
 
+    // TODO MULE-9279 Add generics type once MuleMessage is immutable
     public DataType<?> getDataType()
     {
         return dataType;
     }
 
-    public Object getValue()
+    public T getValue()
     {
         return value;
     }
+
+    // TODO MULE-9279 Remove method once MuleMessage is immutable
+    protected void setDataType(DataType<T> dataType)
+    {
+        this.dataType = dataType;
+    }
+
+    // TODO MULE-9279 Remove method once MuleMessage is immutable
+    protected void setValue(T value)
+    {
+        this.value = value;
+    }
+
+    /**
+     * Performs custom serialization of the value.
+     * <p>
+     * This implementation simply throws a {@link NotSerializableException} as custom serialization is not supported.
+     * Sub-classes can override this method to perform custom 'value' serialization.
+     */
+    protected Serializable serializeValue(ObjectOutputStream out) throws Exception
```

Comment:
I don't think this is correct. This method only executes when the payload is not serializable. I would expect a method with this name to always execute when serialization happens. I think this method should go away and simply throw the exception when the payload is not serializable